solarkraft/src/verify.ts

```diff
@@ -15,6 +11,168 @@ import path from 'node:path'
 const APALACHE_DIST = '/opt/apalache'
 const APALACHE_BIN = path.join(APALACHE_DIST, 'bin', 'apalache-mc')
 
+// TODO(#38): Replace stubs with ITF
+type Value = { type: string; value: any }
+type Binding = { name: string; type: string; value: any }
+type State = Binding[]
+type Transaction = {
+    functionName: string
+    functionArgs: Value[]
+    env: { timestamp: number }
+    error: string
+}
+
+/**
+ * Return an instrumented monitor specification.
+ * @param monitor A TLA+ monitor, as returned from `apalache typecheck --output=temp.json`
+ * @param state The blockchain state before executing the transaction / the monitor.
+ * @param tx The transaction being applied to `state`.
+ * @returns The instrumented TLA+ monitor.
+ */
+export function instrumentMonitor(
+    monitor: any,
+    state: State,
+    tx: Transaction
+): any {
+    // Add a special variable `last_error` that tracks error messages of failed transactions
+    state.push({ name: 'last_error', type: 'TlaStr', value: '' })
+
+    // Declaration of  "Init" (according to `state`)
+    const tlaInit = tlaJsonOperDecl__Conjunction(
+        'Init',
+        state.map((binding) =>
+            tlaJsonEq__NameEx__ValEx(
+                binding.name,
+                false,
+                tlaJsonValEx(binding.type, binding.value)
+            )
+        )
+    )
+
+    // Declaration of "Next" (according to `tx`)
+    const envRecord = tlaJsonRecord([
+        { name: 'timestamp', kind: 'TlaInt', value: tx.env.timestamp },
+    ])
+    const txArgs = tx.functionArgs.map((v) => tlaJsonValEx(v.type, v.value))
+    const tlaNext = tlaJsonOperDecl__Conjunction('Next', [
+        tlaJsonApplication(tx.functionName, [envRecord].concat(txArgs)),
+        tlaJsonEq__NameEx__ValEx(
+            'last_error',
+            true,
+            tlaJsonValEx('TlaStr', tx.error)
+        ),
+    ])
+
+    // Instantiate the monitor spec with declarations of "Init" and "Next"
+    const extendedDeclarations = monitor['modules'][0]['declarations'].concat([
+        tlaInit,
+        tlaNext,
+    ])
+    const extendedModule = {
+        ...monitor['modules'][0],
+        declarations: extendedDeclarations,
+    }
+    return { ...monitor, modules: [extendedModule] }
+}
+
+/**
+ * Return an Apalache TLA+ operator declaration of the form `operDeclName` == `conjuncts`_0 /\ ... /\ `conjuncts`_n.
+ * @param operDeclName Operator name
+ * @param conjuncts Body conjuncts
+ * @returns The operator declaration in Apalache IR JSON: https://apalache.informal.systems/docs/adr/005adr-json.html
+ */
+function tlaJsonOperDecl__Conjunction(
+    operDeclName: string,
+    conjuncts: any
+): any {
+    return {
+        type: 'Untyped',
+        kind: 'TlaOperDecl',
+        name: operDeclName,
+        formalParams: [],
+        isRecursive: false,
+        body: {
+            type: 'Untyped',
+            kind: 'OperEx',
+            oper: 'AND',
+            args: conjuncts,
+        },
+    }
+}
+
```

Comment:
I know that we are still in a prototyping, but would it be good to extract this function and the ones below into a separate module?

Reply:
Also done: 0c9a2b1